doc/design/podgroup-status.md

```diff
@@ -0,0 +1,119 @@
+	// NotEnoughResourcesReason is probed if there're not enough resources to schedule pods
+	NotEnoughResourcesReason string = "NotEnoughResources"
+	// NotEnoughPodsReason is probed if there're not enough tasks compared to `spec.minMember`
+	NotEnoughPodsReason string = "NotEnoughTasks"
+)
+
+// PodGroupState contains details for the current state of this pod group.
+type PodGroupState struct {
+    // Current phase of PodGroup.
+    Phase PodGroupPhase `json:"phase,omitempty" protobuf:"bytes,1,opt,name=phase"`
+	
+    // Last time we probed to this Phase.
+    // +optional
+    LastProbeTime metav1.Time `json:"lastProbeTime,omitempty" protobuf:"bytes,2,opt,name=lastProbeTime"`
+    // Last time the phase transitioned from another to current phase.
+    // +optional
+    LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty" protobuf:"bytes,3,opt,name=lastTransitionTime"`
+    // Unique, one-word, CamelCase reason for the phase's last transition.
+    // +optional
+    Reason string `json:"reason,omitempty" protobuf:"bytes,4,opt,name=reason"`
+    // Human-readable message indicating details about last transition.
+    // +optional
+    Message string `json:"message,omitempty" protobuf:"bytes,5,opt,name=message"`
+}
+
+type PodGroupStatus struct {
+    ......
+
+    // +optional
+    State PodGroupState `json:"state,omitempty" protobuf:"bytes,1,opt,name=state,casttype=State"`
+}
+```
+
+According to the PodGroup's lifecycle, the following phase/state transactions are reasonable. And related
+reasons will be appended to `Reason` field.  
+
+| From          | To            | Reason  |
+|---------------|---------------|---------|
+| Pending       | Running       | When every pods of `spec.minMember` are running |
+| Pending       | Recovering    | When only part of `spec.minMember` are running and the other part pod are rejected by kubelet |
+| Running       | Recovering    | When part of `spec.minMember` have exception, e.g. kill |
+| Recovering    | Running       | When the failed pods re-run successfully |
+| Recovering    | Unschedulable | When the new pod can not be scheduled |
+| Unschedulable | Pending       | When all pods (`spec.minMember`) in PodGroups are deleted |
+| Unschedulable | Running       | When all pods (`spec.minMember`) are deleted |
+
+
+## Feature Interaction
+
+### Cluster AutoScale
+
+[Cluster Autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler) is a tool that
+automatically adjusts the size of the Kubernetes cluster when one of the following conditions is true:
+
+* there are pods that failed to run in the cluster due to insufficient resources,
+* there are nodes in the cluster that have been underutilized for an extended period of time and their pods can be placed on other existing nodes.
+
+For first scenario, the Cluster-Autoscaler need to check PodGroup's `status.State` to know whether there're
+sufficient resources to run the whole group of pods.
```

Comment:
This is not how CA works. It doesn't compute required resources at all. 

Very roughly the way it works is:
0. (Import scheduler predicates from k/k repo)
1. If there are pending pods (pods wit PodScheduled condition set to false with reason given as unschedulable):
1.1 Create an in-memory node object that represents a new node that would result from scale-up.
1.2 Run imported scheduler predicates on pending pods to "schedule" them on in-memory node.
1.3 If some pods remain pending goto 1.1.

This way autoscaler automatically takes into account all scheduling constraints: resources, taints, affinity, storage, etc. Unfortunately it also means it's fundamentally incompatible with any form of custom scheduling. If you want CA to work with kube-batch you need to either encapsulate kube-batch logic in a form of predicates or else duplicate it in CA somehow. It certainly won't be a simple fix. 
It would get even worse for things like gang scheduling. You can swap predicates for something different, but ultimately CA expects a black box that takes (pod, node) pair and tells whether the pod could be scheduled on a given node.

Reply:
What will happen if new node can not be scheduled because of customized plugin? Will it continue to create new nodes?

Reply:
I'm thinking whether CA can keep add new nodes (with some rate & restriction) if `PodGroup` is pending.

Reply:
CA will check what default scheduler would do with the pending pods. If CA thinks the pod can be scheduled on existing nodes it will not add new nodes.

Reply:
That would be hard to work on plugins that can be compatible with CA in the future. In other thread, you mentioned  make pod unschedulable is just tip of an iceber. Do you think we can extract some common criterions in CA for other custom scheduler to meet?  For example, podCondition is just one of them.

Reply:
en. I'm thinking to make CA support as out-of-scope for now; as CA need to work together with default scheduler. We can consider to enhance that in later release.